src/EnergyPlus/PoweredInductionUnits.cc

```diff
@@ -2069,6 +2115,418 @@ void CalcParallelPIU(EnergyPlusData &state,
+    } else if (qdotDelivered1stStageMaxFan < zoneLoad) {
+        thisPIU.heatingOperatingMode = HeatOpModeType::StagedHeatSecondStage;
+        if (thisPIU.UnitType == "AirTerminal:SingleDuct:SeriesPIU:Reheat") {
+            thisPIU.SecAirMassFlow = max(0.0, thisPIU.MaxTotAirMassFlow - primaryAirMassFlow);
+        } else if (thisPIU.UnitType == "AirTerminal:SingleDuct:ParallelPIU:Reheat") {
+            thisPIU.SecAirMassFlow = thisPIU.MaxSecAirMassFlow;
+        }
+    } else if (qdotDelivered1stStageMinFan > zoneLoad) {
+        thisPIU.heatingOperatingMode = HeatOpModeType::HeaterOff;
+        if (thisPIU.UnitType == "AirTerminal:SingleDuct:SeriesPIU:Reheat") {
+            thisPIU.SecAirMassFlow = max(0.0, thisPIU.MinTotAirMassFlow - thisPIU.MinPriAirMassFlow);
+        } else if (thisPIU.UnitType == "AirTerminal:SingleDuct:ParallelPIU:Reheat") {
+            thisPIU.SecAirMassFlow = thisPIU.MinSecAirMassFlow;
+        }
+    }
+}
+
+Real64 CalcVariableSpeedPIUQdotDelivered(EnergyPlusData &state,
+                                         int const PIUNum,   // number of the current PIU being simulated
+                                         int const ZoneNode, // zone node number
+                                         bool const useDAT,
+                                         Real64 const TotAirMassFlow,
+                                         Real64 const fanTurnDown)
+{
+    Real64 qdotDelivered = 0.0;
+    auto &thisPIU = state.dataPowerInductionUnits->PIU(PIUNum);
+    if (thisPIU.UnitType == "AirTerminal:SingleDuct:SeriesPIU:Reheat") {
+        MixerComponent::SimAirMixer(state, thisPIU.MixerName, thisPIU.Mixer_Num);
+        state.dataHVACFan->fanObjs[thisPIU.Fan_Index]->simulate(state, fanTurnDown, _);
+    } else {
+        state.dataHVACFan->fanObjs[thisPIU.Fan_Index]->simulate(state, fanTurnDown, _);
+        MixerComponent::SimAirMixer(state, thisPIU.MixerName, thisPIU.Mixer_Num);
+    }
+    Real64 zoneEnthalpy = Psychrometrics::PsyHFnTdbW(state.dataLoopNodes->Node(ZoneNode).Temp, state.dataLoopNodes->Node(ZoneNode).HumRat);
+    Real64 piuTemp = 0.0;
+    if (useDAT) {
+        piuTemp = thisPIU.designHeatingDAT;
+    } else {
+        piuTemp = state.dataLoopNodes->Node(thisPIU.HCoilInAirNode).Temp;
+    }
+    Real64 piuEnthalpy = Psychrometrics::PsyHFnTdbW(piuTemp, state.dataLoopNodes->Node(ZoneNode).HumRat);
+    qdotDelivered = TotAirMassFlow * (piuEnthalpy - zoneEnthalpy);
+    return qdotDelivered;
+}
+
+void CalcVariableSpeedPIUModulatedHeatingBehavior(EnergyPlusData &state,
+                                                  int const PIUNum,   // number of the current PIU being simulated
+                                                  int const ZoneNode, // zone node number
+                                                  Real64 const zoneLoad,
+                                                  bool const Pri,
+                                                  Real64 const primaryAirMassFlow)
+{
```

Comment:
New function used the determine if, for modulated heating operation, the variable speed fan should run at minimum, maximum, or in between the minimum and maximum to meet the load. If the latter, use regula falsi to find at which fraction it needs to run to meet the load and considers the user-defined design heating discharge air temperature and high limit discharges temperature. Note that this function is by both series and parallel PIUs.